runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java

```diff
@@ -0,0 +1,83 @@
+package org.corfudb.runtime.collections;
+
+import io.netty.buffer.Unpooled;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.util.serializer.ISerializer;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.AbstractMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ *
+ * A wrapper class that makes a RocksIterator compatible with Java {@link Iterator}.
+ *
+ * Created by Maithem on 1/21/20.
+ */
+@NotThreadSafe
+public class RocksDbEntryIterator<K, V> implements Iterator<Map.Entry<K, V>>, AutoCloseable {
+
+    /**
+     * A reference to the underlying rocksdb iterator
+     */
+    final private CheckedRocksIterator checkedIterator;
+
+    /**
+     * Serializer to serialize/deserialize the key/values
+     */
+    final private ISerializer serializer;
+
+    /**
+     * place holder for the current value
+     */
+    private Map.Entry<K, V> next;
+
+    public RocksDbEntryIterator(RocksIterator iterator, ISerializer serializer) {
+        this.checkedIterator = new CheckedRocksIterator(iterator);
+        this.serializer = serializer;
+        iterator.seekToFirst();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasNext() {
+        if (next == null && checkedIterator.isValid()) {
```

Comment:
If we detect that the iterator has reached the end, we should go ahead and relinquish the resources immediately, rather than relying on `try-with-resource` or GC. This is a more robust solution since it is not guaranteed that the client will use `try-with-resource`.

Reply:
Not really, that assumes all stream consumers have to consume the full stream, which is an incorrect assumption. Consumers have to the stream whether explicitly or with try-with-resources